### PR TITLE
add conversation model

### PR DESCRIPTION
**Pull request by cch5ng**
at first I put required on the messages property. but then was thinking maybe after the invitation gets accepted by one person, would this trigger creation of a conversation? then there may be no messages yet.

also how do we differentiate between group chat vs one-on-one chat? like if there are 3 users and you want to have 3 way chat as well as one-one-one chat. does the invitation need to be chat group specific?

### CODE REVIEW

**Comment by rjtbansal** [comment]
> maybe after the invitation gets accepted by one person, would this trigger creation of a conversation?

@cch5ng  From what I understand from the project doc, as soon as invitation gets accepted that person gets added to the sender's 'friends' list. So I think even then conversation need not be initiated. Only when the sender or the receiver initiate the conversation and based on verification whether they are friends only then the conversation should be initiated. Let me know what do you think about this?

> also how do we differentiate between group chat vs one-on-one chat?

I think we could have a separate collection for Group Chats wherein we store ids or email addresses of all those users who are in the group, name of that group chat (maybe even auto-merged names separated by commas just like Slack in case no group chat name is provided) and their array of messages (objects with message text and the sender's id/email). Just thinking aloud here, would like to hear more ideas if what I am proposing makes sense :)

> like if there are 3 users and you want to have 3 way chat as well as one-one-one chat. does the invitation need to be chat group specific?

For this case, to keep it simple, initially I would assume that the person creating a group chat is already 'friends' with other 2 people. If thats the case, then I dont think there needs to be any invitation step involved. All we need to do is create a connection between the 3 people and add them to the group. 
Now since the creator of the group already is friends with others again there wont be an invitation involved. Invitation should only be sent as long as users are not friends, not otherwise from what I understand from project spec. 
Let me know for follow up questions and we can discuss during standup as well.

**Comment by cch5ng** [comment]
ready to review but I think there are dependencies on user collection that may require changes

**Review by rjtbansal — APPROVED** [review]
Looks really good to me @cch5ng so far. Even if there are changes, they can be handled accordingly. For now I think its good to go.

File: server/routes/conversation.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
+const express = require("express");
+const mongoose = require("mongoose");
+const passport = require("passport");
+const Conversation = require("../database/models/conversation");
+const User = require("../database/models/user");
+const router = express.Router();
+
+router.get("/user/:id/conversations",
```
Comment (rjtbansal) [inline server/routes/conversation.js]: I had missed this earlier. Same comment as the other PR. Remove /conversations. This endpoint above due to routes setup will end up looking like:
`/conversations/user/:id/conversations`
After removing /conversations from the end above endpoint looks like:
`/conversations/user/:id`

**Review by rjtbansal — CHANGES_REQUESTED** [review]
Sorry to have missed this in my previous review. Just this minor endpoint change and we should be good to go.

File: server/routes/conversation.js (hunk 1 above)
Comment (cch5ng) [reply server/routes/conversation.js]: done

**Comment by cch5ng** [comment]
@rjtbansal I made the fix to the route. I was thinking that when the user sees conversations, they should be sorted by the last updated date (when last message was added). I'm not that familiar using the .aggregate method and so I added an updated_on date field to the model for now (and sort on that). let me know if you want to review again or if I should just merge.

**Review by rjtbansal — APPROVED** [review]

**Comment by rjtbansal** [comment]
Thats fine @cch5ng . We can look into aggregate later when we have time. This looks good so feel free to merge

**Pull request merged**